### PR TITLE
[SPARK-9408] [PySpark] [MLlib] Refactor linalg.py to /linalg

### PR DESCRIPTION
**Pull request by mengxr**
This is based on @MechCoder 's PR https://github.com/apache/spark/pull/7731. Hopefully it could pass tests. @MechCoder I tried to make minimal changes. If this passes Jenkins, we can merge this one first and then try to move `__init__.py` to `local.py` in a separate PR.

Closes #7731

### CODE REVIEW

File: python/pyspark/sql/types.py [hunk 1]
```diff
@@ -629,7 +629,7 @@ def jsonValue(self):
 
     @classmethod
     def fromJson(cls, json):
-        pyUDT = json["pyClass"]
+        pyUDT = str(json["pyClass"])
```
Comment (mengxr) [inline python/pyspark/sql/types.py]: @davies If I don't add `str(...)`, I got the following exception in `ml/clustering.py`:

~~~
      File "/Users/meng/src/spark/python/lib/pyspark.zip/pyspark/sql/types.py", line 636, in fromJson
        m = __import__(pyModule, globals(), locals(), [pyClass])
    TypeError: ("Item in ``from list'' not a string", <function _parse_datatype_json_string at 0x10436a7d0>, (u'{"type":"struct","fields":[{"name":"features","type":{"type":"udt","class":"org.apache.spark.mllib.linalg.VectorUDT","pyClass":"pyspark.mllib.linalg.VectorUDT","sqlType":{"type":"struct","fields":[{"name":"type","type":"byte","nullable":false,"metadata":{}},{"name":"size","type":"integer","nullable":true,"metadata":{}},{"name":"indices","type":{"type":"array","elementType":"integer","containsNull":false},"nullable":true,"metadata":{}},{"name":"values","type":{"type":"array","elementType":"double","containsNull":false},"nullable":true,"metadata":{}}]}},"nullable":true,"metadata":{}},{"name":"prediction","type":"integer","nullable":true,"metadata":{}}]}',))
~~~

Comment (davies) [inline python/pyspark/sql/types.py]: That's strange, is it possible that you `pyspark.zip` is outdate?

Comment (mengxr) [inline python/pyspark/sql/types.py]: Tried again and saw the same error without `str(...)`. This is the same issue as reported in https://bugs.python.org/issue21720.

Comment (davies) [inline python/pyspark/sql/types.py]: O, I see, LGTM

Comment (davies) [inline python/pyspark/sql/types.py]: Could you add a comment for it?

**Comment by mengxr** [comment]
test this please

**Comment by davies** [comment]
LGTM

**Comment by MechCoder** [comment]
Great, but how were you able to figure this out?

**Comment by mengxr** [comment]
I googled the error message.

**Comment by mengxr** [comment]
test this please

**Comment by mengxr** [comment]
test this please

**Comment by mengxr** [comment]
test this please

**Comment by mengxr** [comment]
test this please

**Comment by mengxr** [comment]
All failed tests are unrelated to this PR. https://github.com/apache/spark/commit/1135551cd59787be37c278be596fd831387d698e passed Jenkins and the the only change in the latest commit is to add a comment to `str(json[...])`. So I'm going to merge this PR to unblock two follow-up PRs.

**Comment by mengxr** [comment]
Merged into master.

**Pull request closed without merging**